bindings.c

```diff
@@ -4092,6 +4292,363 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
+	p->total_pid = total_pid;
+	p->last_pid = last_pid;
+
+	free(line);
+	for (; i > 0; i--)
+		free(idbuf[i-1]);
+out:
+	free(idbuf);
+	return sum;
+}
+/*
+ * Traverse the hash table and update it.
+ */
+void *load_begin(void *arg)
+{
+
+	char *path = NULL;
+	int i, sum, length, ret;
+	struct load_node *f;
+	int first_node;
+	clock_t time1, time2;
+
+	while (1) {
+		time1 = clock();
+		for (i = 0; i < LOAD_SIZE; i++) {
+			pthread_mutex_lock(&load_hash[i].lock);
+			if (load_hash[i].next == NULL) {
+				pthread_mutex_unlock(&load_hash[i].lock);
+				continue;
+			}
+			f = load_hash[i].next;
+			first_node = 1;
+			while (f) {
+				length = strlen(f->cg) + 2;
+				do {
+					/* strlen(f->cg) + '.' or '' + \0 */
+					path = (char *)malloc(length);
+				} while (!path);
+
+				ret = snprintf(path, length, "%s%s", *(f->cg) == '/' ? "." : "", f->cg);
+				if (ret < 0 || ret > length - 1) {
+					/* snprintf failed, ignore the node.*/
+					lxcfs_error("Refresh node %s failed for snprintf().\n", f->cg);
+					goto out;
+				}
+				sum = refresh_load(f, path);
+				if (sum == 0) {
+					f = del_node(f, i);
+				} else {
+out:				f = f->next;
+				}
+				free(path);
+				/* load_hash[i].lock locks only on the first node.*/
+				if (first_node == 1) {
+					first_node = 0;
+					pthread_mutex_unlock(&load_hash[i].lock);
+				}
+			}
+		}
+		time2 = clock();
+		usleep(FLUSH_TIME * 1000000 - (int)((time2 - time1) * 1000000 / CLOCKS_PER_SEC));
+	}
+}
+
+static int proc_loadavg_read(char *buf, size_t size, off_t offset,
+		struct fuse_file_info *fi)
+{
+	struct fuse_context *fc = fuse_get_context();
+	struct file_info *d = (struct file_info *)fi->fh;
+	pid_t initpid;
+	char *cg;
+	size_t total_len = 0;
+	char *cache = d->buf;
+	struct load_node *n;
+	int hash;
+	int cfd;
+	unsigned long a, b, c;
+
+	if (offset) {
+		if (offset > d->size)
+			return -EINVAL;
+		if (!d->cached)
+			return 0;
+		int left = d->size - offset;
+		total_len = left > size ? size : left;
+		memcpy(buf, cache + offset, total_len);
+		return total_len;
+	}
+	if (!loadavg)
+		return read_file("/proc/loadavg", buf, size, d);
+
+	initpid = lookup_initpid_in_store(fc->pid);
+	if (initpid <= 0)
+		initpid = fc->pid;
+	cg = get_pid_cgroup(initpid, "cpu");
+	if (!cg)
+		return read_file("/proc/loadavg", buf, size, d);
+
+	prune_init_slice(cg);
+	hash = calc_hash(cg);
+	n = locate_node(cg, hash);
+
+	/* First time */
+	if (n == NULL) {
+		if (!find_mounted_controller("cpu", &cfd)) {
+			/*
+			 * In locate_node() above, pthread_rwlock_unlock() isn't used
+			 * because delete is not allowed before read has ended.
+			 */
+			pthread_rwlock_unlock(&load_hash[hash].rdlock);
+			return 0;
+		}
+		n = (struct load_node *)malloc(sizeof(struct load_node));
+		if (!n) {
+			lxcfs_error("Create node %s failed for malloc().\n", cg);
+			pthread_rwlock_unlock(&load_hash[hash].rdlock);
+			return read_file("/proc/loadavg", buf, size, d);
+		}
+		do {
+			n->cg = (char *)malloc(strlen(cg)+1);
```

Comment:
Please don't cast the result of `malloc()`.

Reply:
Same question as further above why loop around `malloc()` here and not in other places?